python/variable_init.cpp

```diff
@@ -0,0 +1,456 @@
+
+  // We cannot easily index into shape. Store it to compare to the data later.
+  const bool got_shape = !shape.is_none();
+  auto deduced_shape = got_shape ? shape.cast<std::vector<scipp::index>>()
+                                 : std::vector<scipp::index>(ndim);
+
+  for (scipp::index d = 0; d < ndim; ++d) {
+    scipp::index size = -1;
+    if (values.has_value()) {
+      if (variances.has_value() && values->shape(d) != variances->shape(d)) {
+        throw except::DimensionError(
+            format("The shapes of values and variances differ in dimension ", d,
+                   ": ", py::object(values->attr("shape")), " vs ",
+                   py::object(variances->attr("shape")), '.'));
+      }
+      size = values->shape(d);
+    } else {
+      if (variances.has_value()) {
+        size = variances->shape(d);
+      }
+    }
+    if (got_shape && deduced_shape[d] != size) {
+      throw except::DimensionError(format(
+          "The shape of the data differs from the given shape in dimension ", d,
+          ": ", size, " vs ", deduced_shape[d], '.'));
+    } else {
+      deduced_shape[d] = size;
+    }
+  }
+
+  return Dimensions{dim_labels, deduced_shape};
+}
+
+void ensure_is_scalar(const py::buffer &array) {
+  if (const auto ndim = array.attr("ndim").cast<int64_t>(); ndim != 0) {
+    throw except::DimensionError(
+        format("Cannot interpret ", ndim, "-dimensional array as a scalar."));
+  }
+}
+
+template <class T>
+T extract_scalar(const py::object &obj, const units::Unit unit) {
+  using TM = ElementTypeMap<T>;
+  using PyType = typename TM::PyType;
+  TM::check_assignable(obj, unit);
+  if (py::isinstance<py::buffer>(obj)) {
+    ensure_is_scalar(obj);
+    return converting_cast<PyType>::cast(obj.attr("item")());
+  } else {
+    return converting_cast<PyType>::cast(obj);
+  }
+}
+
+template <>
+core::time_point extract_scalar<core::time_point>(const py::object &obj,
+                                                  const units::Unit unit) {
+  using TM = ElementTypeMap<core::time_point>;
+  using PyType = typename TM::PyType;
+  TM::check_assignable(obj, unit);
+  if (py::isinstance<py::buffer>(obj)) {
+    ensure_is_scalar(obj);
+    return core::time_point{obj.attr("astype")(py::dtype::of<PyType>())
+                                .attr("item")()
+                                .template cast<PyType>()};
+  } else {
+    return core::time_point{obj.cast<PyType>()};
+  }
+}
+
+template <>
+python::PyObject extract_scalar<python::PyObject>(const py::object &obj,
+                                                  const units::Unit unit) {
+  using TM = ElementTypeMap<python::PyObject>;
+  TM::check_assignable(obj, unit);
+  return obj;
+}
+
+template <class T> struct MakeVariableDefaultInit {
+  static Variable apply(const Dimensions &dims, const units::Unit unit,
+                        const std::optional<bool> with_variance) {
+    auto var = with_variance.value_or(false)
```

Comment:
very minor
```suggestion
    auto var = bool(with_variance)
```

Reply:
Is that really the same? Even if it is, it would be much less readable, since it convoluted the optional with its value.

Reply:
Yes correct. I misread this first time it's doing the right thing as it is.